tools/codegen/pkg/schemapatch/schemapatch.go

```diff
@@ -0,0 +1,88 @@
+package schemapatch
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/openshift/api/tools/codegen/pkg/generation"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/schemapatcher"
+)
+
+const openshiftFeatureSetEnv = "OPENSHIFT_REQUIRED_FEATURESET"
+
+// executeSchemaPatchForGroupVersionWithBinary runs a schemapatch on the controller-gen binary against the group version
+// provided. If any requiredFeatureSets are present it will set the appropriate environment variable to ensure the
+// generator only executes the generator on the correct features sets.
+func executeSchemaPatchForGroupVersionWithBinary(controllerGen string, group string, version generation.APIVersionContext, versionPaths []string, requiredFeatureSets sets.String) error {
+	if requiredFeatureSets.Len() > 0 {
+		// The controller generator picks up feature sets from an env var.
+		if err := os.Setenv(openshiftFeatureSetEnv, strings.Join(requiredFeatureSets.List(), ",")); err != nil {
+			return fmt.Errorf("could not set env %s: %w", openshiftFeatureSetEnv, err)
+		}
+
+		defer os.Unsetenv(openshiftFeatureSetEnv)
+	}
+
+	args := []string{}
+
+	args = append(args, manifestsArg(version.Path))
+	args = append(args, pathsArg(versionPaths))
+	args = append(args, outputArg(version.Path))
+
+	klog.V(2).Infof("Generating API schema for %s/%s", group, version.Name)
+	cmd := exec.Command(controllerGen, args...)
+
+	if err := cmd.Run(); err != nil {
```

Comment:
Will this show stdout/stderr from external controller-gen?

Reply:
Good catch, it wouldn't, I have updated it so that now it will :)